hazelcast-client/src/main/java/com/hazelcast/client/txn/ClientTransactionManager.java

```diff
@@ -175,4 +206,46 @@ public void shutdown() {
         recoveredTransactions.clear();
     }
 
+    private Address getRandomAddress() {
+
+        MemberImpl member = (MemberImpl) loadBalancer.next();
+        if (member == null) {
+            Set<Member> members = client.getCluster().getMembers();
+            String msg;
+            if (members.isEmpty()) {
+                msg = "No address was return by the LoadBalancer since there are no members in the cluster";
+            } else {
+                msg = "No address was return by the LoadBalancer. "
+                        + "But the cluster contains the following members:" + members;
+            }
+            throw new IllegalStateException(msg);
+        }
+
+        return member.getAddress();
+    }
+
+    private class ClusterAuthenticator implements Authenticator {
```

Comment:
I am making ClusterAuthenticator a separate class to remove the code duplication.